samples/nrf9160/gps/src/supl_support.c

```diff
@@ -0,0 +1,154 @@
+#include <zephyr.h>
+#include <supl_session.h>
+#include <stdio.h>
+#include <logging/log.h>
+
+#define SUPL_SERVER "supl.google.com"
+
+static int            supl_fd;
+
+int open_supl_socket(void)
+{
+	int err;
+	int proto;
+	u16_t port;
+	struct nrf_addrinfo *addr;
+	struct nrf_addrinfo *info;
+
+	proto = NRF_IPPROTO_TCP;
+	port = nrf_htons(7276);
+
+	struct nrf_addrinfo hints = {
+		.ai_family = NRF_AF_INET,
+		.ai_socktype = NRF_SOCK_STREAM,
+		.ai_protocol = proto,
+		/* Either a valid,
+		 * NULL-terminated access point name or NULL.
+		 */
+		.ai_canonname = NULL,
+	};
+
+	err = nrf_getaddrinfo(SUPL_SERVER, NULL, &hints, &info);
+	if (err) {
+		printk("Failed to resolve hostname %s on IPv4, errno: %d)\n",
+			log_strdup(SUPL_SERVER),
+			errno);
+		return -1;
+	}
+
+	supl_fd = nrf_socket(NRF_AF_INET, NRF_SOCK_STREAM, proto);
+	if (supl_fd < 0) {
+		printk("Failed to create socket, errno %d\n", errno);
+		goto cleanup;
+	}
+
+	struct nrf_timeval timeout = {
+		.tv_sec = 1,
+		.tv_usec = 0,
+	};
+
+	err = nrf_setsockopt(supl_fd,
+			     NRF_SOL_SOCKET,
+			     NRF_SO_RCVTIMEO,
+			     &timeout,
+			     sizeof(timeout));
+	if (err) {
+		printk("Failed to setup socket timeout, errno %d\n", errno);
+		return -1;
+	}
+
+	/* Not connected */
+	err = -1;
+
+	for (addr = info; addr != NULL; addr = addr->ai_next) {
+		struct nrf_sockaddr *const sa = addr->ai_addr;
+
+		switch (sa->sa_family) {
+		case NRF_AF_INET6:
+			((struct nrf_sockaddr_in6 *)sa)->sin6_port = port;
+			break;
+		case NRF_AF_INET:
+			((struct nrf_sockaddr_in *)sa)->sin_port = port;
+			char ip[255] = { 0 };
+
+			nrf_inet_ntop(NRF_AF_INET,
+				      (void *)&((struct nrf_sockaddr_in *)
+					      sa)->sin_addr,
+				      ip,
+				      255);
+			printf("ip %s (%x) port %d\n",
+				log_strdup(ip),
+				((struct nrf_sockaddr_in *)sa)->sin_addr.s_addr,
+				nrf_ntohs(port));
+			break;
+		}
+
+		err = nrf_connect(supl_fd, sa, addr->ai_addrlen);
+		if (err) {
+			/* Try next address */
+			printk("Unable to connect, errno %d\n", errno);
+		} else {
+			/* Connected */
+			break;
+		}
+	}
+
+cleanup:
+	nrf_freeaddrinfo(info);
+
+	if (err) {
+		/* Unable to connect, close socket */
+		nrf_close(supl_fd);
+		supl_fd = -1;
+	}
+
+	return 0;
+}
+
+void close_supl_socket(void)
+{
+	if (nrf_close(supl_fd) < 0) {
+		printk("Failed to close SUPL socket\n");
+	}
+}
+
+ssize_t supl_write(const void *p_buff, size_t nbytes, void *user_data)
+{
+	ARG_UNUSED(user_data);
+	return nrf_send(supl_fd, p_buff, nbytes, 0);
+}
+
+int logger(int level, const char *fmt, ...)
+{
+	char buffer[CONFIG_LOG_STRDUP_MAX_STRING] = { 0 };
+	va_list args;
+
+	va_start(args, fmt);
+	int ret = vsnprintk(buffer, sizeof(buffer), fmt, args);
+
+	va_end(args);
+
+	if (ret < 0) {
+		printk("%s: encoding error\n", __func__);
```

Comment:
If this function is called directly, `__func__` will always resolve to `logger`.
If it is not called directly, but wrapped in a macro, I would not put `buffer` on the stack since it affects all the functions that log something.

Reply:
It's called through the api->logger function pointer by the SUPL client. Maybe I should call the function supl_logger or something to make less likely that this is confused by the real logger.